### PR TITLE
Gradual activation quantization in Keras

### PR DESCRIPTION
**Pull request by reuvenperetz**
## Pull Request Description:
Add support for gradual activation quantization in Keras.
This mainly converts torch implementation of gradual activation quantization to a common implementation.
In addition, pytests tests were added to keras (similar to current pytorch tests).

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/gptq/common/regularization_factory.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
+# Copyright 2024 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/regularization_factory.py:1]: why? this is not a new file

File: model_compression_toolkit/trainable_infrastructure/common/annealing_schedulers.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        if not (0 <= t_start < t_end):
+            raise ValueError(f'Expected 0 <= t_start < t_end, actual {t_end=} {t_start=}')
+
+        self.t_start = t_start
+        self.t_end = t_end
+        self.initial_val = initial_val
+        self.target_val = target_val
+
+    def _compute_factor(self, t: int) -> float:
```
Comment (ofirgo) [inline model_compression_toolkit/trainable_infrastructure/common/annealing_schedulers.py:37]: is it possible to use `@abstractmethod` here? (not sure because it does not inherit from ABC)

File: model_compression_toolkit/trainable_infrastructure/common/annealing_schedulers.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        Calculates the annealed value based on the current time step `t`.
+
+        Args:
+            t: Current time step.
+
+        Returns:
+            float: Annealed value between initial_val and target_val.
+        """
+        factor = self._compute_factor(t)
```
Comment (ofirgo) [inline model_compression_toolkit/trainable_infrastructure/common/annealing_schedulers.py:62]: nice! I wish we had more of these in our code

File: model_compression_toolkit/gptq/keras/quantizer/quantization_builder.py [hunk 4]
```diff
@@ -70,12 +73,13 @@ def quantization_builder(n: common.BaseNode,
...
+                                                        quantizer_id=TrainingMethod.STE,
                                                         quant_method=quant_method,
-                                                        quantizer_base_class=BaseKerasInferableQuantizer)
+                                                        quantizer_base_class=BaseKerasActivationTrainableQuantizer)
+        cfg = get_trainable_quantizer_activation_config(n, None)
 
-        kwargs = get_inferable_quantizer_kwargs(n.final_activation_quantization_cfg, QuantizationTarget.Activation)
-
-        activation_quantizers.append(quantizer_class(**kwargs))
+        # freeze_quant_params is True since in GPTQ the quantization parameters should not be trained (previously, we used inferable quantizers).
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/keras/quantizer/quantization_builder.py:82]: "the **activation** quantization parameters should not..."
and you can remove the "previously.." part IMO

File: model_compression_toolkit/gptq/keras/quantization_facade.py [hunk 5]
```diff
@@ -94,19 +96,37 @@ def get_keras_gptq_config(n_epochs: int,
...
             >>> gptq_conf = mct.gptq.get_keras_gptq_config(n_epochs=5)
 
             Other Tensorflow optimizers can be passed:
 
             >>> gptq_conf = mct.gptq.get_keras_gptq_config(n_epochs=3, optimizer=tf.keras.optimizers.Nadam())
 
-            The configuration can be passed to :func:`~model_compression_toolkit.keras_post_training_quantization` in order to quantize a keras model using gptq.
+            To enable Gradual Activation Quantization with non-default settings build GradualActivationQuantizationConfig:
+
+            >>> gradual_act_conf = mct.gptq.GradualActivationQuantizationConfig(mct.gptq.QFractionLinearAnnealingConfig(initial_q_fraction=0.2))
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/keras/quantization_facade.py:109]: this looks very scary, do we really need to specify `initial_q_fraction`?
 even providing the `GradualActivationQuantizationConfig` and not use a default one seems like to much understanding of the algorithm to ask for from the user

File: model_compression_toolkit/gptq/keras/quantization_facade.py [hunk 6]
```diff
@@ -94,19 +96,37 @@ def get_keras_gptq_config(n_epochs: int,
...
+            >>> gptq_conf = mct.gptq.get_keras_gptq_config(n_epochs=3, gradual_activation_quantization=gradual_act_conf)
+
+            The configuration can be passed to :func:`~model_compression_toolkit.keras_gradient_post_training_quantization` in order to quantize a keras model using gptq.
+
 
         """
+        optimizer = optimizer or tf.keras.optimizers.Adam(learning_rate=LR_DEFAULT)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/keras/quantization_facade.py:116]: If we are touching this API, lets make sure that this is aligned with the arguments and values of the parallel pytorch API (it might be fine but just lets be sure)

File: model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py [hunk 7]
```diff
@@ -29,6 +29,7 @@
 from model_compression_toolkit import constants as C
 
 from model_compression_toolkit.qat.keras.quantizer.base_keras_qat_weight_quantizer import BaseKerasQATWeightTrainableQuantizer
+
```
Comment (ofirgo) [inline model_compression_toolkit/qat/keras/quantizer/lsq/symmetric_lsq.py:32]: unnecessary

File: model_compression_toolkit/gptq/keras/gptq_training.py [hunk 8]
```diff
@@ -77,6 +81,15 @@ def __init__(self,
             hessian_info_service: HessianScoresService for fetching and computing Hessian's approximation scores.
 
         """
+
+        def _get_total_grad_steps():
+            return get_total_grad_steps(representative_data_gen) * gptq_config.n_epochs
+
+        # must be set prior to model building in the base class constructor. TODO: why?
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/keras/gptq_training.py:88]: why really?

Comment (irenaby) [reply model_compression_toolkit/gptq/keras/gptq_training.py:88]: it is used in activation holder construction so it must be set before the model is built, which happens in the base constructor

File: tests_pytest/keras/gptq/test_gradual_act_quantization.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class TestGradualActivationQuantization:
+
+    def test_gradual_act_quant_wrapper(self, x):
+        quantizer = Quantizer()
+        qw = GradualActivationQuantizerWrapper(quantizer, q_fraction_scheduler=lambda t: t / (t + 1))
+
+        y0, y1, y2 = [qw(x, training=True) for _ in range(3)]
+        np.testing.assert_array_almost_equal(y0.numpy(), x.numpy())  # t=0
```
Comment (ofirgo) [inline tests_pytest/keras/gptq/test_gradual_act_quantization.py:45]: Why np.testing? is there any advantages of using this tool?
unless there is a clear advantage I think it is preferred that we'll be aligned with the assertion methodology that Irena used for the Pytorch Pytests (which is, to use Python's built in assertion).

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/gptq/keras/quantization_facade.py (hunk 5 above)
Comment (reuvenperetz) [reply model_compression_toolkit/gptq/keras/quantization_facade.py:109]: It's possible to remove this, but we need to demonstrate to the user how to use and configure GradualActivationQuantizationConfig somehow.

Do you think this should be in the GPTQ tutorial instead? Even though I think it should be here since this is what will be generated in the docs pages.

Comment (ofirgo) [reply model_compression_toolkit/gptq/keras/quantization_facade.py:109]: I  do think it should be in a tutorial, and here we should only have a "plug-and-play" example. I think that just giving a line of code that configures something without getting into the meaning of every argument and module that is used can be confusing

File: model_compression_toolkit/gptq/keras/quantization_facade.py (hunk 6 above)
Comment (ofirgo) [reply model_compression_toolkit/gptq/keras/quantization_facade.py:116]: I have no idea how the values were chosen and why they differ between the frameworks. Let's align both defaults to the default value in the accuracy runner (I think its 1e-2 for both lr and reg factor but double check that)

File: tests_pytest/keras/__init__.py [hunk 10]
```diff
@@ -1,14 +1,14 @@
-# Copyright 2024 Sony Semiconductor Israel, Inc. All rights reserved.
+#  Copyright 2024 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (ofirgo) [inline tests_pytest/keras/__init__.py:1]: remove spaces

**Review by ofirgo — APPROVED** [review]
